internal/app/upstream/client_test.go

```diff
@@ -0,0 +1,393 @@
+	cancel()
+}
+
+func TestOpenStreamShouldSendTheResponseOnTheChannel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	responseChan := make(chan *v2.DiscoveryResponse)
+	response := &v2.DiscoveryResponse{}
+	client := GetMockClient(ctx, CallOptions{Timeout: time.Second}, nil, responseChan, func(m interface{}) error {
+		responseChan <- response
+		return nil
+	})
+
+	resp, err := client.OpenStream(ctx, &v2.DiscoveryRequest{
+		TypeUrl: listenerTypeURL,
+		Node:    &core.Node{},
+	})
+	assert.Nil(t, err)
+	assert.NotNil(t, resp)
+	val := <-resp
+	assert.Equal(t, val.Response, response)
+	assert.Nil(t, val.Err)
+	cancel()
+}
+
+func TestOpenStreamShouldSendTheNextRequestWithUpdatedVersionAndNonce(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	responseChan := make(chan *v2.DiscoveryResponse)
+	lastAppliedVersion := ""
+	index := 0
+	client := GetMockClient(ctx, CallOptions{Timeout: time.Second}, nil, responseChan, func(m interface{}) error {
+		message := m.(*v2.DiscoveryRequest)
+
+		assert.Equal(t, message.GetVersionInfo(), lastAppliedVersion)
+		assert.Equal(t, message.GetResponseNonce(), lastAppliedVersion)
+
+		response := &v2.DiscoveryResponse{
+			VersionInfo: strconv.Itoa(index),
+			Nonce:       strconv.Itoa(index),
+			TypeUrl:     listenerTypeURL,
+		}
+		lastAppliedVersion = strconv.Itoa(index)
+		index++
+		responseChan <- response
+		return nil
+	})
+
+	resp, err := client.OpenStream(ctx, &v2.DiscoveryRequest{
+		TypeUrl: listenerTypeURL,
+		Node:    &core.Node{},
+	})
+	assert.Nil(t, err)
+	assert.NotNil(t, resp)
+	for i := 0; i < 5; i++ {
+		val := <-resp
+		assert.Equal(t, val.Response.GetVersionInfo(), strconv.Itoa(i))
+		assert.Equal(t, val.Response.GetNonce(), strconv.Itoa(i))
+		assert.Nil(t, val.Err)
+	}
+
+	cancel()
+}
+
+func TestOpenStreamShouldSendErrorWhenSendMsgBlocks(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	responseChan := make(chan *v2.DiscoveryResponse)
+	client := GetMockClient(ctx, CallOptions{Timeout: time.Nanosecond}, nil, responseChan, func(m interface{}) error {
+		<-ctx.Done()
+		return nil
+	})
+
+	resp, err := client.OpenStream(ctx, &v2.DiscoveryRequest{
+		TypeUrl: listenerTypeURL,
+		Node:    &core.Node{},
+	})
+	assert.Nil(t, err)
+	assert.NotNil(t, resp)
+	val := <-resp
+	assert.Equal(t, val.Err.Error(), "rpc error: code = DeadlineExceeded desc = timeout")
+	assert.Nil(t, val.Response)
+	cancel()
+}
+
+func GetMockClient(
+	ctx context.Context,
+	callOptions CallOptions,
+	errorOnCreate error,
+	receiveChan chan *v2.DiscoveryResponse,
+	sendCb func(m interface{}) error) Client {
+	return NewMockClient(
+		ctx,
+		createMockLdsClient(errorOnCreate, receiveChan, sendCb),
+		createMockRdsClient(errorOnCreate, receiveChan, sendCb),
+		createMockEdsClient(errorOnCreate, receiveChan, sendCb),
+		createMockCdsClient(errorOnCreate, receiveChan, sendCb),
+		callOptions,
+	)
+}
+
+func createMockLdsClient(
```

Comment:
Rather than doing this, I think creating an internal/ or test/utils/fake package with mock xDS clients would be useful unless this is the only module requiring mocked out xDS clients.

Reply:
I didn't take the approach since xdsclient mock is used only in one place.

Reply:
I think orchestrator could use these rather than reimplementing. These will be valuable to be shareable imo.

Reply:
I think the mock needed by orchestrator is bit different. It wouldn't need the complex setup provided by this particular mock in discussion. The mock needed by orchestrator would only need a simple `OpenStream` mock for easy testing. WDYT?

Reply:
I have moved the mock into a reusable package. LMK how does it look.